cardano-integration/src/Test/Integration/Cardano/Local.hs

```diff
@@ -0,0 +1,444 @@
+
+      byronGenesisDir <- createByronGenesis workspace startTime testnetMagic options
+      shelleyGenesisDir <- createShelleyGenesisStaked workspace testnetMagic options
+
+      let
+        wallets = [1..numWallets] <&> \n -> PaymentKeyPair
+          { paymentSKey = shelleyGenesisDir </> "utxos" </> "utxo" <> show n <> ".skey"
+          , paymentVKey = shelleyGenesisDir </> "utxos" </> "utxo" <> show n <> ".vkey"
+          }
+
+        delegators = [1..numDelegators] <&> \n -> Delegator
+          { paymentKeyPair = PaymentKeyPair
+            { paymentSKey = shelleyGenesisDir </> "stake-delegator-keys" </> "payment" <> show n <> ".skey"
+            , paymentVKey = shelleyGenesisDir </> "stake-delegator-keys" </> "payment" <> show n <> ".vkey"
+            }
+          , stakingKeyPair = StakingKeyPair
+            { stakingSKey = shelleyGenesisDir </> "stake-delegator-keys" </> "staking" <> show n <> ".skey"
+            , stakingVKey = shelleyGenesisDir </> "stake-delegator-keys" </> "staking" <> show n <> ".vkey"
+            }
+          }
+
+      network <- setupNetwork workspace byronGenesisDir shelleyGenesisDir
+      spoNodes <- traverse (setupSpoNode workspace options network logsDir socketDir byronGenesisDir shelleyGenesisDir) [1..numSpoNodes]
+
+      liftIO $ mapConcurrently_ assertChainExtended spoNodes
+
+      pure LocalTestnet{..}
+    \e -> do
+      -- prevent workspace from being cleaned up for diagnosing errors.
+      _ <- unprotect $ W.releaseKey workspace
+      throw e
+
+createByronGenesis :: MonadIO m => Workspace -> UTCTime -> Int -> LocalTestnetOptions -> m FilePath
+createByronGenesis workspace startTime testnetMagic LocalTestnetOptions{..} = do
+  byronGenesisSpecFile <- writeWorkspaceFileJSON workspace "byron.genesis.spec.json" $ object
+    [ "heavyDelThd"       .= ("300000000000" :: String)
+    , "maxBlockSize"      .= ("2000000" :: String)
+    , "maxTxSize"         .= ("4096" :: String)
+    , "maxHeaderSize"     .= ("2000000" :: String)
+    , "maxProposalSize"   .= ("700" :: String)
+    , "mpcThd"            .= ("20000000000000" :: String)
+    , "scriptVersion"     .= (0 :: Int)
+    , "slotDuration"      .= show slotDuration
+    , "unlockStakeEpoch"  .= ("18446744073709551615" :: String)
+    , "updateImplicit"    .= ("10000" :: String)
+    , "updateProposalThd" .= ("100000000000000" :: String)
+    , "updateVoteThd"     .= ("1000000000000" :: String)
+    , "softforkRule"      .= object
+      [ "initThd"           .= ("900000000000000" :: String)
+      , "minThd"            .= ("600000000000000" :: String)
+      , "thdDecrement"      .= ("50000000000000" :: String)
+      ]
+    , "txFeePolicy"       .= object
+      [ "multiplier"        .= ("43946000000" :: String)
+      , "summand"           .= ("155381000000000" :: String)
+      ]
+    ]
+  let byronGenesisDir = resolveWorkspacePath workspace "byron-genesis"
+  execCli_
```

Comment:
We could just call the function in the `cardano-cli` package, instead of exec'ing it.

Reply:
True, we could. I'll look into this. I guess this way, we don't need to depend on `cardano-cli` at the cabal level.